process.go

```diff
@@ -0,0 +1,483 @@
+	if err != nil {
+		return err
+	}
+
+	modifyRequestStr := string(modifyRequestb)
+
+	err = hcsModifyProcess(process.handle, modifyRequestStr, &resultp)
+	err = processHcsResult(err, resultp)
+	if err != nil {
+		err = &ProcessError{Operation: operation, Process: process, Err: err}
+		logrus.Error(err)
+		return err
+	}
+
+	logrus.Debugf(title+" succeeded processid=%d", process.processID)
+	return nil
+}
+
+// Close cleans up any state associated with the process but does not kill
+// or wait on it.
+func (process *process) Close() error {
+	operation := "Close"
+	title := "HCSShim::Process::" + operation
+	logrus.Debugf(title+" processid=%d", process.processID)
+
+	// Don't double free this
+	if process.handle == 0 {
+		return nil
+	}
+
+	if err := hcsCloseProcess(process.handle); err != nil {
+		err = &ProcessError{Operation: operation, Process: process, Err: err}
+		logrus.Error(err)
+		return err
+	}
+
+	process.handle = 0
+
+	logrus.Debugf(title+" succeeded processid=%d", process.processID)
+	return nil
+}
+
+// closeProcess wraps process.Close for use by a finalizer
+func closeProcess(process *process) {
+	process.Close()
+}
+
+func (process *process) registerCallback(expectedNotification hcsNotification) (uintptr, error) {
+	var (
+		callbackHandle hcsCallback
+		channel        notificationChannel = make(chan error, 1)
+	)
+	processCallbackMapLock.Lock()
+	defer processCallbackMapLock.Unlock()
+
+	callbackNumber := nextProcessCallback
+	nextProcessCallback++
+
+	context := &notifcationWatcherContext{
+		expectedNotification: expectedNotification,
+		channel:              channel,
+	}
+	processCallbackMap[callbackNumber] = context
+
+	err := hcsRegisterProcessCallback(process.handle, processNotificationWatcherCallback, callbackNumber, &callbackHandle)
+	if err != nil {
+		return 0, err
+	}
+	context.handle = callbackHandle
+
+	return callbackNumber, nil
+}
+
+func (process *process) unregisterCallback(callbackNumber uintptr) error {
+	processCallbackMapLock.Lock()
+	defer processCallbackMapLock.Unlock()
+	handle := processCallbackMap[callbackNumber].handle
+
+	if handle == 0 {
+		return nil
+	}
+
+	err := hcsUnregisterProcessCallback(handle)
+	if err != nil {
+		return err
+	}
+
+	processCallbackMap[callbackNumber] = nil
+
+	handle = 0
+
+	return nil
+}
+
+func (process *process) waitForNotificationTimeout(callbackNumber uintptr, timeout time.Duration) error {
+	processCallbackMapLock.Lock()
+	channel := processCallbackMap[callbackNumber].channel
+	processCallbackMapLock.Unlock()
+
+	select {
+	case err := <-channel:
+		return err
+	case <-time.After(timeout):
+		return ErrTimeout
+	}
+}
+
+func (process *process) waitForNotification(callbackNumber uintptr) error {
+	processCallbackMapLock.Lock()
+	channel := processCallbackMap[callbackNumber].channel
+	processCallbackMapLock.Unlock()
+
+	select {
+	case err := <-channel:
+		return err
+	}
+}
+
+func (e *ProcessError) Error() string {
+	if e == nil {
```

Comment:
I'm not sure it can, but https://golang.org/src/net/net.go?s=11560:11681#L407 makes sure it is not nil. I'm leaning towards removing this, but unsure because of the check in net.